hyperspy/signal.py

```diff
@@ -3288,99 +3288,227 @@ def save(
+            )
+
+        # Deprecation warning for extension parameter
+        if extension is not None:
+            warnings.warn(
+                "The 'extension' parameter is deprecated in HyperSpy 2.4 and will be removed in HyperSpy 3.0. "
+                "Please use 'file_format' instead.",
+                FutureWarning,
+                stacklevel=2,
+            )
+
         if filename is None:
             if self.tmp_parameters.has_item(
                 "filename"
             ) and self.tmp_parameters.has_item("folder"):
+                # Determine the extension to use
+                if self.tmp_parameters.has_item("extension"):
+                    file_ext = self.tmp_parameters.extension
+                elif file_format is not None:
+                    # Get the default extension for the file format from rsciio
+                    try:
+                        from hyperspy.io import _format_name_to_reader
+
+                        writer = _format_name_to_reader(file_format)
+                        file_ext = (
+                            "." + writer["file_extensions"][writer["default_extension"]]
+                        )
+                    except (ValueError, KeyError):
+                        # If format not found in rsciio, use the file_format as extension
+                        file_ext = "." + file_format
+                else:
+                    file_ext = ".hspy"  # Default extension
+
                 filename = Path(
-                    self.tmp_parameters.folder, self.tmp_parameters.filename
-                )
-                extension = (
-                    self.tmp_parameters.extension if not extension else extension
+                    self.tmp_parameters.folder,
+                    self.tmp_parameters.filename + file_ext,
                 )
+                # Don't override extension if it was explicitly provided
+                if extension is None and self.tmp_parameters.has_item("extension"):
+                    extension = self.tmp_parameters.extension
             elif self.metadata.has_item("General.original_filename"):
                 filename = self.metadata.General.original_filename
             else:
                 raise ValueError("File name not defined")
 
         if not isinstance(filename, MutableMapping):
             filename = Path(filename)
-            if extension is not None:
-                filename = filename.with_suffix(f".{extension}")
+
+            # Check if filename is clearly a directory path.
+            # We only consider it a directory path if:
+            # 1. It's an existing directory, OR
+            # 2. The path explicitly ends with a directory separator ('/' or '\')
+            # This conservative approach ensures we don't accidentally treat
+            # filenames without extensions as directories.
+            is_directory_path = filename.is_dir() or str(filename).endswith(("/", "\\"))
```

Comment:
`filename.is_dir()` will returns False if the directory doesn't exist. What is the expectation here?